tensorflow/python/util/compat.py

```diff
@@ -40,6 +40,7 @@
 import numpy as _np
 import six as _six
 
+from tensorflow.python.framework import tensor_shape
```

Comment:
I don't like this dependency. util should be small utilities which do not depend on any of the concrete tensorflow stuff.

I think the other fix is better.

Reply:
Thanks for comment. Could you please give me a bit more detail about where to fix?

How about hack [`def _FilterInt(v):`](https://github.com/tensorflow/tensorflow/blob/c48aaadbe9c5802316d73e40bbcde34221277b81/tensorflow/python/framework/tensor_util.py#L238) function? 
or add a `if` branch at [`_AssertCompatible(values, dtype)`](https://github.com/tensorflow/tensorflow/blob/c48aaadbe9c5802316d73e40bbcde34221277b81/tensorflow/python/framework/tensor_util.py#L383)?

Reply:
Essentially, I do not think that this approach is the right way to fix the problem. compat.as_int shouldn't know anything about dimension. I think something like the change made in  #12558  is the right way to go.

I'll close this pull request now.

Reply:
Hi, @alextp . Perhaps the [new commit](https://github.com/facaiy/tensorflow/commits/BUG/tf_reshape_mix_int_with_dim) will be interesting for you. Unfortunately, the PR has been closed.

Reply:
How about #12127 ? It is more straightforward.